tracer/tracer.go

```diff
@@ -0,0 +1,94 @@
+package tracer
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+
+	lightstep "github.com/lightstep/lightstep-tracer-go"
+	opentracing "github.com/opentracing/opentracing-go"
+	jaeger "github.com/uber/jaeger-client-go"
+	"github.com/uber/jaeger-client-go/transport"
+	"github.com/uber/jaeger-client-go/transport/udp"
+	"github.com/uber/jaeger-client-go/transport/zipkin"
+)
+
+const (
+	backendJaeger    = "jaeger"
+	backendLightstep = "lightstep"
+)
+
+// NewTracer returns a new opentracing.Tracer via the options provided.
+// If this isn't called the default tracer is opentracaing.NoopTracer{},
+// resulting in no tracing being captured.
+func NewTracer(opts Opts) (opentracing.Tracer, error) {
+	switch opts.Backend {
+	case backendJaeger:
+		return newJaegerTracer(opts)
+	case backendLightstep:
+		return newLightstepTracer(opts)
+	}
+
+	return nil, fmt.Errorf("unknown tracer backend: '%s'", opts.Backend)
+}
+
+func newJaegerTracer(opts Opts) (opentracing.Tracer, error) {
+	// Detect whether we're using zipkin-compatible tracing or jaeger's UDP
+	// remote.
+	u, err := url.Parse(opts.Addr)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing tracing-addr: %s", err)
+	}
+
+	var trnsprt transport.Transport
+	if strings.Contains(u.Scheme, "http") {
+		trnsprt, err = zipkin.NewHTTPTransport(opts.Addr)
+	} else {
+		// no http or https, therefore we must be using jaeger's
+		// reporter via UDP using its default max packet length.
+		trnsprt, err = udp.NewUDPTransport(opts.Addr, 0)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("error configuring tracer: %s", err)
+	}
+
+	// jaeger has many samplers available, ranging from constant sampling
+	// to guaranteed throughput probabilistic sampling. For ease of
+	// configuration, by default use constant sampling.
+	//
+	// TODO(tonyhb): close the io.Closer provided by jaeger when a node is
+	// terminated to drain the span queue
+	tracer, _ := jaeger.NewTracer(opts.Hostname, jaeger.NewConstSampler(true), jaeger.NewRemoteReporter(trnsprt))
+	return tracer, nil
+}
+
+func newLightstepTracer(opts Opts) (opentracing.Tracer, error) {
+	return lightstep.NewTracer(lightstep.Options{
+		AccessToken: opts.Addr,
+	}), nil
+}
+
+// Opts hold all configuration opts for creating a new opentracing.Tracer via
+// NewTracer
+type Opts struct {
+	// Backend is the type of tracer used, either jaeger or lightstep
+	Backend string
+
+	// Addr is used to reference either the zipkin backend URL for
+	// collection; the jaeger UDP host and port for collection; or the
+	// lightstep access token for collection
```

Comment:
Minor point, but it's ultimately going to be confusing to overload "addr" in this way... e.g., LightStep's tracer also supports a host:port for non-SaaS collectors. My suggestion would be to pass in the entire jaeger XOR lightstep as a single structured string (json, whatever)... gross, but won't turn into a proliferation of flags for Docker binaries.